Step05 (#10)
WebAhead/authentication-workshop: walkthrough/step05.md

        return res.send('Passwords do not match')
    }
    
    // else continue to the next middleware 
    next()
    
}

app.post('/create-user', validateNewUser, (req, res, next) => {

    // Add new user to our database
    // and end request    
});

```

In this example the `validateNewUser` middleware is the second argument for `app.post`, you can have as many functions or middlewares as you want. for example:

```javascript=

app.post('/create-user', validateNewUser, anotherMiddleware, andAnother, (req, res, next) => {

    // Add new user to our database
    // and end request    
});

```

Or you can have them in an array:
 
```javascript=
app.post('/create-user', [validateNewUser, anotherMiddleware, andAnother], (req, res, next) => {

    // Add new user to our database
    // and end request    
});

```

**important**: all of the middlewares above need to either run the `next` method or end the request. so your server does not hang.

Middlewares are very powerful when your servers get really big and you don't want to repeat yourself, it is totally valid and even encouraged to have a folder just for your middlewares.

# Tasks:

#### 1. build the auth middleware

> Let's use what we learned about middlewares above to implement an auth middleware that checks if there is actual cookie and if it's valid.

Reviewer: ```suggestion
Let's use what we learned about middlewares above to implement an auth middleware that checks if there is an actual cookie and if it's valid.
```

- we have a middlewares folder with an index file, create an `authCheck` file that exports a function.
- in the `authCheck` file take the code from `home.js` that checks for the cookie and validates it.
> Note: you do not want to render the home page if the cookie is valid or not valid in the middleware, you need to let our home controller handle the page rendering.
- if everything is fine and valid, set the username to the username that was stored in the cookie and set signedIn to true using [res.locals](https://expressjs.com/en/4x/api.html#res.locals).
-  if he is not logged in set username to be null and signedIn to be false, after all that run the `next` method to go to the next middleware or controller.
- then in the `middlewares/index.js` folder import `authCheck` and export it within an object.

> - and then add that middleware before our home controller. using the [example of route specific middleware above](#An-example-of-route-specific-middleware).

Reviewer: ```suggestion
- Add that middleware before our home controller. using the [example of route specific middleware above](#An-example-of-route-specific-middleware).
```

> - From there [fetch our username variable](http://expressjs.com/en/api.html#app.get) and signedIn variable and pass it to our hbs page.

Reviewer: ```suggestion
- [Fetch our username variable](http://expressjs.com/en/api.html#app.get) and signedIn variable and pass it to our hbs page.
```

- if an error occured also pass it in the `res.locals` so the home page can handle the proper rendering.

> We should have the same functionality as the last step but this time we can easly reuse our authCheck before any route.

Reviewer: ```suggestion
We should have the same functionality as the last step but this time we can easily reuse our authCheck before any route.
```

#### 2. Implement login time middleware

We want to track when the user logged in to the website, by logging the date and time in which he logged it

(https://www.npmjs.com/package/winston#quick-start).
- create middleware specifically for the `authenticate` route that apppends the time and date (and the username) in which the user has logged in to a log file.
- You can either use Node's `fs` module to write the log files or for an extra challenge try to use [Winston]

> - the middleware can run before or after the logout controller your decision.

Reviewer: ```suggestion
- It is up to you whether to run the middleware before or after the logout controller.
```

# Step05 - Auth middleware

Now that we have our user passwords secured using bcrypt, and our cookies encoded with JWT, we can safley say that we took a step forward into having a secured website.

There are many more steps that we can take to secure our website properly(links would be available at the bottom), but for now this is a good basis.

<img src="" alt="" styles="text-align:center;" />

> **insert security related image here**

Reviewer: Can't find one

Author: i'll just put a row of security related icon

Right now the only place where we check if the user is logged in and valid is in the homepage, what if we wanted to check if he is logged in 10 different other routes ?

> So do we write the same code multple time ? No cause then it is not considered to be DRY code (Don't Repeat Yourself)

Reviewer: ```suggestion
Should we write the same code for different routes? 
**No** because then it is not DRY (Don't Repeat Yourself) code, remember that you should be encapsulating code when you can :sparkles: 
```

That's where Express middleware comes in, we had a brief introduction about middleware last week using:
- body-parser
- express static handler
- compression
- logger middleware
- etc.

### What are middlewares ?

> middlewares functions are functions that have access to the request object (req), the response object (res), and the next middleware function in the application’s request-response cycle.

Reviewer: ```suggestion
Middleware functions are functions that have access to the request object (req), the response object (res), and the next middleware function in the application’s request-response cycle.
```

});

```

In this example the `validateNewUser` middleware is the second argument for `app.post`, you can have as many functions or middlewares as you want. for example:

```javascript=

app.post('/create-user', validateNewUser, anotherMiddleware, andAnother, (req, res, next) => {

    // Add new user to our database
    // and end request    
});

```

Or you can have them in an array:
 
```javascript=
app.post('/create-user', [validateNewUser, anotherMiddleware, andAnother], (req, res, next) => {

    // Add new user to our database
    // and end request    
});

```

**important**: all of the middlewares above need to either run the `next` method or end the request. so your server does not hang.

Middlewares are very powerful when your servers get really big and you don't want to repeat yourself, it is totally valid and even encouraged to have a folder just for your middlewares.

# Tasks:

#### 1. build the auth middleware

Let's use what we learned about middlewares above to implement an auth middleware that checks if there is actual cookie and if it's valid.

- we have a middlewares folder with an index file, create an `authCheck` file that exports a function.

> - in the `authCheck` function take the code from `home.js` that checks for the cookie and validates it.

Reviewer: authCheck is both a file and the function exported?\

Author: well the file only exports a function, so technically both the function and file are `authCheck` but i can see how this would be confusing, so i'll change
